modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/writer/ColumnValueWriter.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.batch.engine.internal.writer;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.lang.reflect.Field;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * @author Ivica cardic
+ */
+public class ColumnValueWriter {
+
+	public void write(Object item, Consumer<List<?>> consumer)
+		throws IllegalAccessException {
+
+		Map<String, Object> columnNameValueMap = _getColumnNameValueMap(item);
+
+		Set<Map.Entry<String, Object>> entries = columnNameValueMap.entrySet();
+
+		Stream<Map.Entry<String, Object>> stream = entries.stream();
+
+		columnNameValueMap = stream.filter(
+			entry -> _isFieldEligibleForWrite(item, entry.getKey())
+		).flatMap(
+			this::_getEntryStreams
+		).collect(
+			Collectors.toMap(
+				Map.Entry::getKey,
+				entry -> {
+					if (entry.getValue() == null) {
+						return StringPool.BLANK;
+					}
+
+					return entry.getValue();
+				})
+		);
+
+		List<String> columnNames = new ArrayList<>(columnNameValueMap.keySet());
```

Comment:
@ivicac I think it is a bug here. In case of Map value, how do you know all items that are writing out having the same Map keys? If you just pick the first item's map's keyset for the headers, there is a good chance some following items have mismatched map's keyset. Then this entire output will be wrong.

Reply:
hi @shuyangzhou, yes, you are right. But I set the assumption that we will have the same maps as handling maps is mostly because of localization. You will notice that I don't handle deep graph (serializing nested complex objects), too, as for all those cases there is JSON or JSONL format

Reply:
also, can we close this PR and review this one: https://github.com/shuyangzhou/liferay-portal/pull/8431. It includes commits, above. If you remember you said, you want to see usage, and also wanted me to leave streams, as you have already made changes.(I had to make some slight change in order to get it to work, still)

Reply:
@shuyangzhou to leave things simple, I will add a check that will throw an exception if there is a different number of columns in any row.
It will be in the next PR.

Reply:
@shuyangzhou I changed my mind. I added the ability to optionally define field names you want to export. That means if there is such a case then API consumer can explicitly define filed names he wants to include to export.
It is a  part of the next PR/s